From 3ef29954c03afc9d172ed34135202cb6793de840 Mon Sep 17 00:00:00 2001
From: Lim Siew Hoon <siew.hoon.lim@intel.com>
Date: Mon, 31 Jul 2023 18:42:26 +0800
Subject: [PATCH 43/77] video: Add new GST_VIDEO_FORMAT_DMA_DRM video format

This GST_VIDEO_FORMAT_DMA_DRM is introduced for DMABuf kind feature
usage. It represent the DMA DRM kind memory. And like the ENCODED
format, it should not be interpreted and mapped as normal video format.

Part-of: <https://gitlab.freedesktop.org/gstreamer/gstreamer/-/merge_requests/4981>

Backport from upstream:
https://gitlab.freedesktop.org/gstreamer/gstreamer/-/commit/94c94d34dc4b0d5f539c3ae98305136e98e81ef8

Signed-off-by: Lim Siew Hoon <siew.hoon.lim@intel.com>
---
 gst-libs/gst/video/video-converter.c |  1 +
 gst-libs/gst/video/video-format.c    |  8 ++++++--
 gst-libs/gst/video/video-format.h    | 15 +++++++++++++++
 gst-libs/gst/video/video-info.c      |  1 +
 4 files changed, 23 insertions(+), 2 deletions(-)

diff --git a/gst-libs/gst/video/video-converter.c b/gst-libs/gst/video/video-converter.c
index 84a91976..08937348 100644
--- a/gst-libs/gst/video/video-converter.c
+++ b/gst-libs/gst/video/video-converter.c
@@ -7358,6 +7358,7 @@ get_scale_format (GstVideoFormat format, gint plane)
     case GST_VIDEO_FORMAT_Y412_LE:
     case GST_VIDEO_FORMAT_NV12_8L128:
     case GST_VIDEO_FORMAT_NV12_10BE_8L128:
+    case GST_VIDEO_FORMAT_DMA_DRM:
       res = format;
       g_assert_not_reached ();
       break;
diff --git a/gst-libs/gst/video/video-format.c b/gst-libs/gst/video/video-format.c
index d869f9a1..84b6be37 100644
--- a/gst-libs/gst/video/video-format.c
+++ b/gst-libs/gst/video/video-format.c
@@ -7293,6 +7293,8 @@ static const VideoFormat formats[] = {
   MAKE_YUV_T_FORMAT (NV12_10BE_8L128, "raw video",
       GST_MAKE_FOURCC ('N', 'T', '1', '2'), DPTH10_10_10, PSTR0, PLANE011,
       OFFS001, SUB420, PACK_NV12_10BE_8L128, TILE_10bit_8x128 (LINEAR)),
+  {0x00000000, {GST_VIDEO_FORMAT_DMA_DRM, "DMA_DRM", "DMA DRM video",
+          GST_VIDEO_FORMAT_FLAG_COMPLEX, DPTH0, PSTR0, PLANE_NA, OFFS0}},
 };
 
 G_GNUC_END_IGNORE_DEPRECATIONS;
@@ -7768,7 +7770,8 @@ generate_raw_video_formats (gpointer data)
 
     all->formats[i] = gst_video_format_from_string (g_value_get_string (v));
     g_assert (all->formats[i] != GST_VIDEO_FORMAT_UNKNOWN
-        && all->formats[i] != GST_VIDEO_FORMAT_ENCODED);
+        && all->formats[i] != GST_VIDEO_FORMAT_ENCODED
+        && all->formats[i] != GST_VIDEO_FORMAT_DMA_DRM);
   }
 
   g_value_unset (&list);
@@ -7855,7 +7858,8 @@ gst_video_make_raw_caps_with_features (const GstVideoFormat formats[],
       GValue v = G_VALUE_INIT;
 
       g_return_val_if_fail (formats[i] != GST_VIDEO_FORMAT_UNKNOWN
-          && formats[i] != GST_VIDEO_FORMAT_ENCODED, NULL);
+          && formats[i] != GST_VIDEO_FORMAT_ENCODED
+          && formats[i] != GST_VIDEO_FORMAT_DMA_DRM, NULL);
 
       g_value_init (&v, G_TYPE_STRING);
       g_value_set_static_string (&v, gst_video_format_to_string (formats[i]));
diff --git a/gst-libs/gst/video/video-format.h b/gst-libs/gst/video/video-format.h
index 7a645cca..86155357 100644
--- a/gst-libs/gst/video/video-format.h
+++ b/gst-libs/gst/video/video-format.h
@@ -145,6 +145,10 @@ G_BEGIN_DECLS
  * @GST_VIDEO_FORMAT_NV12_16L32S: NV12 with 16x32 Y tiles and 16x16 UV tiles. (Since: 1.22)
  * @GST_VIDEO_FORMAT_NV12_8L128 : NV12 with 8x128 tiles in linear order (Since: 1.22)
  * @GST_VIDEO_FORMAT_NV12_10BE_8L128 : NV12 10bit big endian with 8x128 tiles in linear order (Since: 1.22)
+ * @GST_VIDEO_FORMAT_DMA_DRM: DMA DRM special format. It's only used with
+ *                            memory:DMABuf #GstCapsFeatures, where an extra
+ *                            parameter (drm-format) is required to define the
+ *                            image format and its memory layout.
  *
  * Enum value describing the most common video formats.
  *
@@ -400,6 +404,17 @@ typedef enum {
    * Since: 1.22
    */
   GST_VIDEO_FORMAT_NV12_10BE_8L128,
+  /**
+   * GST_VIDEO_FORMAT_DMA_DRM:
+   *
+   * @GST_VIDEO_FORMAT_DMA_DRM represent the DMA DRM special format. It's
+   * only used with memory:DMABuf #GstCapsFeatures, where an extra
+   * parameter (drm-format) is required to define the image format and
+   * its memory layout.
+   *
+   * Since: 1.24
+   */
+  GST_VIDEO_FORMAT_DMA_DRM,
 } GstVideoFormat;
 
 #define GST_VIDEO_MAX_PLANES 4
diff --git a/gst-libs/gst/video/video-info.c b/gst-libs/gst/video/video-info.c
index 03a2f477..3552c5ae 100644
--- a/gst-libs/gst/video/video-info.c
+++ b/gst-libs/gst/video/video-info.c
@@ -1267,6 +1267,7 @@ fill_planes (GstVideoInfo * info, gsize plane_size[GST_VIDEO_MAX_PLANES])
       break;
     }
     case GST_VIDEO_FORMAT_ENCODED:
+    case GST_VIDEO_FORMAT_DMA_DRM:
       break;
     case GST_VIDEO_FORMAT_UNKNOWN:
       GST_ERROR ("invalid format");
-- 
2.40.1

